bin/improver-extract

```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# -----------------------------------------------------------------------------
+# (C) British Crown Copyright 2018 Met Office.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Script to extract subcubes given constraints."""
+
+from improver.argparser import ArgParser
+from improver.utilities.load import load_cube
+from improver.utilities.cube_extraction import (parse_constraint_list,
+                                                extract_subcube)
+from improver.utilities.save import save_netcdf
+
+
+def main():
+    """Invoke data extraction."""
+
+    parser = ArgParser(description='Extracts subset of data from a single '
+                       'input file, subject to equality-based constraints.')
+    parser.add_argument('input_file', metavar='INPUT_FILE',
+                        help="File containing a dataset to extract from.")
+    parser.add_argument('output_file', metavar='OUTPUT_FILE',
+                        help="File to write the extracted dataset to.")
+    parser.add_argument('constraints', metavar='CONSTRAINTS', nargs='+',
```

Comment:
Yes, I think that there are a few options for handling the constraints here. I guess the benefit of this approach is that the e.g. `threshold=0.1` syntax matches Iris' approach, although there might be other options e.g. `threshold, 0.1, mm/hr name, probability_of_precipitation` or something else, which might mean that you wouldn't need to specify the None value for units that you don't want to change. However, as I don't have a clear suggestion, the current approach is probably sufficient.